[TFLite] Add int16x8 support for BATCH_MATMUL operator

Tessil opened the pull request:
Hi,

This PR adds int16x8 support for the BATCH_MATMUL operator in TensorFlow Lite.

Thibaut

Tessil committed:
Add int16x8 support for BATCH_MATMUL operator

Tessil commented on tensorflow/lite/toco/tflite/op_version.cc:
```diff
@@ -66,6 +66,7 @@ std::string GetMinimumRuntimeVersionForModel(const Model& model) {
           {{OperatorType::kBatchToSpaceND, 1}, "1.6.0"},
           {{OperatorType::kBatchToSpaceND, 2}, "1.14.0"},
           {{OperatorType::kBatchMatMul, 1}, kPendingReleaseOpVersion},
+          {{OperatorType::kBatchMatMul, 3}, kPendingReleaseOpVersion},
```
The version 2 of the operator seems to be missing but it's present in `runtime_version.cc`. Should I add it in this PR?

The `kPendingReleaseOpVersion` also seems to be outdated compared to `runtime_version.cc`.

talumbau replied:
Line 30 here seems to indicate that no additional updates are required to this file. I would revert this change. The change in `tensorflow/lite/tools/versioning/op_version.cc` seems sufficient. thanks!

Tessil replied:
Thanks I saw the comment on line 30 but I wasn't completely sure if we still needed to update it as it may still be used by TOCO when experimental_new_converter is manually set to false in the TFLiteConverter. I removed the line.

talumbau replied:
OK great, thanks! I'll now try to get this submitted.

Tessil replied:
In the end we don't need to update this file at all any more when bumping the version of an operator? I did it for the following merged PR https://github.com/tensorflow/tensorflow/pull/41487, but I will update my other non-merged PRs to remove the update of `tensorflow/lite/toco/tflite/op_version.cc` then.

Tessil commented on tensorflow/lite/kernels/batch_matmul_test.cc:
```diff
@@ -553,6 +558,34 @@ TEST_P(QuantizedBatchMatMulOpTest, SimpleTestQuantizedInt8) {
+  QuantizedBatchMatMulOpModel m(
+      /*units=*/3, /*batches*/ 2,
+      /*lhs=*/
+      {TensorType_INT16, {2, 10}, 0, 0, inputs_scale, 0},
```
I currently set the input scale and zero-point manually as it seems that the `QuantizationParams` method in `test_utils.h` called by `AddTensor` returns a zero-point of -1 when I pass a symmetric range but the int16x8 BATCH_MATMUL kernel is restricted to a zero-point of 0. 

For example calling `QuantizationParams<int16_t>(-10.0, 10.0)` returns a -1 zero-point, is it an expected behaviour?

Tessil committed:
Fix the inputs_scale in the test

talumbau commented on tensorflow/lite/kernels/batch_matmul.cc:
```diff
@@ -382,9 +400,14 @@ TfLiteStatus TransposeRowsColumns(TfLiteContext* context,
-    TF_LITE_KERNEL_LOG(context,
-                       "Can only transpose tensors with float and int8 type.");
+    TF_LITE_KERNEL_LOG(
+        context, "Can only transpose tensors with float, int8 and int16 type.");
```
The original comment wasn't as clear as it could be. I think it should be `float, int8 or int16 type`

Tessil replied:
Thank you for the review, the `or` is fffectively more appropriate. Made the change.

talumbau reviewed (changes requested):
Thanks for this change! Looks good.

Tessil committed:
Clarify transpose unsupported tensor type error message

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/16x8_batch_matmul

Tessil committed:
Update BATCH_MATMUL operator version supported by the reference kernel

Tessil committed:
Merge branch 'master' into toupstream/16x8_batch_matmul

Tessil committed:
All the inputs of the BATCH_MATMUL should be quantized by the type of the activation

talumbau reviewed (commented):
This looks good. Just one small change. thanks for the PR!

Tessil committed:
Remove unnecessary versioning of kBatchMatMul in deprecated file

Tessil committed:
Fix compilation error, lhs should be used and not rhs

Tessil commented:
@talumbau  I introduced a small error when fixing a previous merge conflict, I fixed it in the latest commit. The PR will need re-approval, sorry for that.

talumbau reviewed (changes requested):
Thanks for these changes. I had a question about a change in the tests, but overall looking good.

Tessil commented on tensorflow/lite/kernels/batch_matmul_test.cc:
```diff
@@ -483,7 +483,12 @@ class QuantizedBatchMatMulOpModel : public SingleOpModel {
-    rhs_id_ = AddInput({lhs.type, {input_size_, units_}, lhs.min, lhs.max});
+    rhs_id_ = AddInput({lhs.type,
+                        {input_size_, units_},
+                        0,
+                        0,
```
The `lhs_id_ = AddInput(lhs);` line will create an input tensor with `lhs_id_` and set the scaling and zero-point based either on the min/max range of the parameter `t` or on the provided scaling and zero_point of parameter `t` if the range is left to 0 (see [SingleOpModel::AddTensor](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/test_util.h#L626)). The change I did on the creation of the `rhs_id_` input allows to directly recuperate the calculated scaling and zero-point of `lhs_id_` without needing any recalculation if a min/max range is used and is also more general as it allows the `lhs` parameter of `QuantizedBatchMatMulOpModel` to either have a min/max range or an explicit scaling and zero point as done in `SimpleTestQuantizedInt16`.

I set the input scale and zero-point manually in `SimpleTestQuantizedInt16` instead of letting `SingleOpModel::AddTensor` do it based on the range to ensure that the zero-point is null (int16 tensors are restricted to null zero-points). The problem is that even if I pass a symmetric range like [-10, 10], the scaling is calculated on the  asymmetric [2^15, 2^15 - 1] range and the zero-point will thus be -1. If needed I could change the test a bit to use a kMin/kMax multiplication factor to nudge a bit the range similar to what was done for some other int16 tests.

talumbau replied:
Yep, I see what you are saying. Looks good! 👍

Tessil committed:
Merge branch 'master' into toupstream/16x8_batch_matmul

Tessil commented:
I merged the latest change from the master branch into the PR as there was an [error in the CI](https://source.cloud.google.com/results/invocations/87d483b1-e13e-434c-8fc3-f8417e716340/targets/%2F%2Ftensorflow%2Fcore:ops_tests/log) coming from a bug in the master that was fixed by 9fe68d5108d8e4aefd7907489cf649da81097516. 

@talumbau the PR will need re-approval, thanks!

Tessil commented:
@gbaned It seems the PR was indirectly merged with the previous approval. Should I just close the PR?

gbaned commented:
@Tessil  I will close it. 
Seems auto-merge is not happening but the changes are now committed, so we can close this. Thank you for the PR.
